packages/at_stream/lib/src/at_notification_stream_channel.dart

```diff
@@ -0,0 +1,351 @@
+  /// side that you are disconnecting and it can too.
+  Future<NotificationResult> sendDisconnectMessage() {
+    AtKey key = AtKey()
+      ..key = "${_messageCounter++}.$sessionId"
+      ..namespace = "$domainNamespace.$baseNamespace"
+      ..sharedWith = otherAtsign
+      ..sharedBy = atClient.getCurrentAtSign()!
+      ..metadata = (Metadata()
+        ..isEncrypted = true
+        ..namespaceAware = false);
+    return notify(
+      key,
+      jsonEncode({"control": "disconnect"}),
+      checkForFinalDeliveryStatus: true,
+      waitForFinalDeliveryStatus: false,
+      ttln: const Duration(hours: 1),
+    );
+  }
+
+  Future close() {
+    // Even though sendDisconnectMessage is already called inside the
+    // sendController.listen's onDone, the onDone function isn't
+    // called before close completes, thus we potentially run it twice
+    // If a client calls close then exits, then notification is never sent.
+    // Thus, we send disconnect first to ensure that the notification stream
+    // is closed and free up the other side's resources sooner.
+    timer?.cancel();
+    return sendDisconnectMessage().then((_) => _sendController.close());
+  }
+
+  void onTimeout() {
+    logger.shout("Session $sessionId timed out.");
+    close();
+  }
+
+  void resetTimer() {
+    timer?.cancel();
+    timer = Timer(sessionTimeout, onTimeout);
+  }
+
+  @protected
+  @visibleForTesting
+  Future<void> initBase({FutureOr<void> Function()? onDone}) async {
+    timer = Timer(sessionTimeout, onTimeout);
+    // sending
+    _sendController.stream.transform(sendTransformer).listen(
+      (Encoded value) async {
+        resetTimer();
+        logger.info("Sending message: $value");
+
+        AtKey key = AtKey()
+          ..key = "${_messageCounter++}.$sessionId"
+          ..namespace = "$domainNamespace.$baseNamespace"
+          ..sharedWith = otherAtsign
+          ..sharedBy = atClient.getCurrentAtSign()!
+          ..metadata = (Metadata()
+            ..isEncrypted = true
+            ..namespaceAware = false);
+        try {
+          final data = {'payload': value};
+          await notify(
+            key,
+            jsonEncode(data),
+            checkForFinalDeliveryStatus: true,
+            waitForFinalDeliveryStatus: false,
+            ttln: const Duration(hours: 1),
+          );
+        } catch (e) {
+          logger.severe("Failed to send message: $e");
+        }
+      },
+      onDone: () async {
+        await onDone?.call();
+        await _recvController.close();
+      },
+    );
+
+    // receiving
+    subscribe(
+      regex: "$sessionId\\.$domainNamespace\\.$baseNamespace$otherAtsign",
+      shouldDecrypt: true,
+    )
+        .transform(NotificationTransformer<Encoded>(this))
+        .transform(recvTransformer)
+        .listen((value) => _recvController.add(value));
+  }
+}
+
+@visibleForTesting
+class NotificationTransformer<Encoded>
```

Comment:
This transformer is applied to the receiving stream before the user's transformer is. It doesn't manipulate any Notifications, but it does remove notifications which are control messages. It provides a Dart record of (notification, value).

- notification provided for advanced use cases.
- value provided since the notification.value is a custom json structure that stores the user's value along side other information.